File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
push = heapq.heappush
pop = heapq.heappop

def solution(jobs):
    pq = []
    rst = 0
    cur_time = 0
    cur_jobs = 0
    jobs.sort()
    def _push():
        nonlocal cur_jobs
        push(pq, (jobs[cur_jobs][1], jobs[cur_jobs][0], cur_jobs))
        cur_jobs += 1
    _push()
    while pq:
        t, req_time, _ = pop(pq)
        cur_time = max(cur_time, req_time)
        cur_time += t
        rst += cur_time - req_time
        while cur_jobs < len(jobs) and jobs[cur_jobs][0] <= cur_time:
            _push()
        if not pq and cur_jobs < len(jobs):
            _push()
    return(rst // len(jobs))
        

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <iostream>

using namespace std;

enum{
    START, DURATION
};

struct cmp{
    bool operator()(vector<int> &a, vector<int> &b){
        return a[DURATION] > b[DURATION];
    }
};

int solution(vector<vector<int>> jobs) {
    int answer = 0;
    int i = 0;
    int time = 0;
    sort(jobs.begin(), jobs.end());
    priority_queue<vector<int>, vector<vector<int>>, cmp> q;
    while(i < jobs.size() || !q.empty()){
        if(i < jobs.size() &&time >= jobs[i][START])
            q.push(jobs[i++]);
        else if(!q.empty()){
            answer += q.top()[DURATION] + time - q.top()[START];
            time += q.top()[DURATION];
            q.pop();
        }
        else
            time = jobs[i][START];
    }
    return answer / jobs.size();
}